File: Sean的leetcode/python/Valid Parentheses.py
###
# Time Complexity: O(n)
# Space Complexity: O(n)
# Warning: check stack is empty in the end
###
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s:
            return True
        d = {'(': ')', '{': '}', '[': ']'}
        stk = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stk.append(c)
            else:
                if not stk or d[stk.pop()] != c:
                    return False
        return not bool(stk) 

File: MyLeetCode/Leetcode/Valid Parentheses.cpp
class Solution {
public:   
    bool isValid(string s) {
        unordered_map<char, char> table = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        stack<char> st;
        for (auto c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            }
            else if (c == ')' || c == '}' || c == ']') {
                if (st.empty()) return false;
                char tmp = st.top();
                st.pop();
                if (table[tmp] != c) {
                    return false;
                }
            }
        }
        return st.empty();
    }
};

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '{' || s[i] == '[' || s[i] == '(') {
               st.push(s[i]); 
            }
            else if (!st.empty() && ((s[i] == '}' && st.top() == '{') || (s[i] == ']' && st.top() == '[') || (s[i] == ')' && st.top() == '('))) {
                st.pop();
            }
            else{
                return false;
            }
        }
        return st.empty();
    }
};